### PR TITLE
feat: add workspace build timing metrics

### PR DESCRIPTION
**Pull request by kevinh-canva**
# Context

We want to place a tight SLO around coder workspace build times, so we can detect regression. However, buffy GPU instances often take a much longer time to start/stop than general instance, which frequently triggered our SLO alerts, even though it's only because of a few (expected) slow GPU builds. This is caused by the metrics we are using `coderd_provisionerd_job_timing_seconds` not having a dimension for template name (as we have a separate template for GPU and another for general instances).

Looking closer at the code, this metrics is also not the correct one to use either, because a Job can actually be many different things, not just a workspace build.

# Intent

This PR introduces a new prometheus metrics for `workspace_build_timing_seconds`, which specifically reports workspace build times. To reduce cardinality, this metrics excludes `workspace_name` and `workspace_owner` that are present on the `workspace_builds_total` metrics. 

This'd allow us to have different (and tight) SLOs for each of our template (GPU vs non-GPU) by filtering on the `template_name` (optionally `template_version` tag) as well as the workspace transition (as we noticed `stop` is often slower than `start`, but users don't care a lot about `stop` transitions).

### CODE REVIEW

**Comment by kevinh-canva** [comment]
I have read the CLA Document and I hereby sign the CLA

**Review by dannykopping — COMMENTED** [review]
Thanks for your contribution @kevinh-canva!

Let's see if we can find a solution to the potential cardinality explosion.

File: provisionerd/provisionerd.go [hunk 1]
```diff
@@ -178,6 +178,22 @@ func NewMetrics(reg prometheus.Registerer) Metrics {
 				Name:      "workspace_builds_total",
 				Help:      "The number of workspaces started, updated, or deleted.",
 			}, []string{"workspace_owner", "workspace_name", "template_name", "template_version", "workspace_transition", "status"}),
+			WorkspaceBuildTimings: auto.NewHistogramVec(prometheus.HistogramOpts{
+				Namespace: "coderd",
+				Subsystem: "provisionerd",
+				Name:      "workspace_build_timings_seconds",
+				Help:      "workspace build time in seconds.",
```
Comment (dannykopping) [inline provisionerd/provisionerd.go:185]: Nit:
```suggestion
				Help:      "The time it takes for a workspace build to complete.",
```

**Review by dannykopping — APPROVED** [review]
LGTM

@kylecarbs could you please force-merge this PR?
The two failing CI jobs are both related to forks not being able to access secrets.

**Comment by dannykopping** [comment]
> I have read the CLA Document and I hereby sign the CLA

@kevinh-canva would you mind commenting this again? The CLA step should now be fixed.

**Comment by kevinh-canva** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by kevinh-canva** [comment]
@dannykopping Looks like it's still failing for me somehow

**Comment by deansheather** [comment]
Sorry about the issues with the CLA, we've been having some troubles lately with secrets in our GitHub actions workflows

**Pull request merged**